File: backend/src/test/java/io/chesslave/model/PieceTest.java
package io.chesslave.model;

import static org.hamcrest.CoreMatchers.hasItem;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

import javaslang.collection.Set;
import org.junit.Test;

public class PieceTest {

    @Test
    public void pieceTypeValues() {
        assertEquals(1, Piece.Type.PAWN.value());
        assertEquals(3, Piece.Type.KNIGHT.value());
        assertEquals(3, Piece.Type.BISHOP.value());
        assertEquals(5, Piece.Type.ROOK.value());
        assertEquals(9, Piece.Type.QUEEN.value());
        assertEquals(0, Piece.Type.KING.value());
    }

    @Test
    public void isFriendTest() {
        final Piece whiteKing = Piece.of(Piece.Type.KING, Color.WHITE);
        final Piece whiteQueen = Piece.of(Piece.Type.QUEEN, Color.WHITE);
        final Piece blackKing = Piece.of(Piece.Type.KING, Color.BLACK);
        assertTrue(whiteKing.isFriend(whiteQueen));
        assertFalse(whiteKing.isFriend(blackKing));
    }

    @Test
    public void isOpponentTest() {
        final Piece whiteKing = Piece.of(Piece.Type.KING, Color.WHITE);
        final Piece whiteQueen = Piece.of(Piece.Type.QUEEN, Color.WHITE);
        final Piece blackKing = Piece.of(Piece.Type.KING, Color.BLACK);
        assertTrue(whiteKing.isOpponent(blackKing));
        assertFalse(whiteKing.isOpponent(whiteQueen));
    }

    @Test
    public void allTest() {
        final Set<Piece> allPieces = Piece.all();
        assertEquals(12, allPieces.size());
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.KING, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.QUEEN, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.ROOK, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.BISHOP, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.KNIGHT, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.PAWN, Color.WHITE)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.KING, Color.BLACK)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.QUEEN, Color.BLACK)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.ROOK, Color.BLACK)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.BISHOP, Color.BLACK)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.KNIGHT, Color.BLACK)));
        assertThat(allPieces, hasItem(Piece.of(Piece.Type.PAWN, Color.BLACK)));
    }

    @Test
    public void equalsTest() {
        final Piece whiteKing = Piece.of(Piece.Type.KING, Color.WHITE);
        assertTrue(whiteKing.equals(whiteKing));
        assertFalse(whiteKing.equals(new Object()));
        assertTrue(whiteKing.equals(Piece.of(Piece.Type.KING, Color.WHITE)));
    }

    @Test
    public void hashCodeTest() {
        final Piece whiteKing = Piece.of(Piece.Type.KING, Color.WHITE);
        final int whiteKingHash = whiteKing.hashCode();
        assertEquals(whiteKingHash, whiteKing.hashCode());
        assertNotEquals(whiteKingHash, new Object().hashCode());
        assertEquals(whiteKingHash, Piece.of(Piece.Type.KING, Color.WHITE).hashCode());
    }
}

File: backend/src/test/java/io/chesslave/model/PieceTest.kt
package io.chesslave.model

import io.chesslave.model.Piece.Type
import org.hamcrest.CoreMatchers.hasItem
import org.junit.Assert.*
import org.junit.Test

class ColorTest {

    @Test
    fun opponentTest() {
        assertEquals(Color.BLACK, Color.WHITE.opponent())
        assertEquals(Color.WHITE, Color.BLACK.opponent())
    }
}

class PieceTest {

    @Test
    fun pieceTypeValues() {
        assertEquals(1, Type.PAWN.value)
        assertEquals(3, Type.KNIGHT.value)
        assertEquals(3, Type.BISHOP.value)
        assertEquals(5, Type.ROOK.value)
        assertEquals(9, Type.QUEEN.value)
        assertEquals(0, Type.KING.value)
    }

    @Test
    fun isFriendTest() {
        assertTrue(Piece.whiteKing.isFriend(Piece.whiteQueen))
        assertFalse(Piece.whiteKing.isFriend(Piece.blackKing))
    }

    @Test
    fun isOpponentTest() {
        assertTrue(Piece.whiteKing.isOpponent(Piece.blackKing))
        assertFalse(Piece.whiteKing.isOpponent(Piece.whiteQueen))
    }

    @Test
    fun allTest() {
        val allPieces = Piece.all
        assertEquals(12, allPieces.size())
        assertThat(allPieces, hasItem(Piece.whiteKing))
        assertThat(allPieces, hasItem(Piece.whiteQueen))
        assertThat(allPieces, hasItem(Piece.whiteRook))
        assertThat(allPieces, hasItem(Piece.whiteBishop))
        assertThat(allPieces, hasItem(Piece.whiteKnight))
        assertThat(allPieces, hasItem(Piece.whitePawn))
        assertThat(allPieces, hasItem(Piece.blackKing))
        assertThat(allPieces, hasItem(Piece.blackQueen))
        assertThat(allPieces, hasItem(Piece.blackRook))
        assertThat(allPieces, hasItem(Piece.blackBishop))
        assertThat(allPieces, hasItem(Piece.blackKnight))
        assertThat(allPieces, hasItem(Piece.blackPawn))
    }

    @Test
    fun equalsTest() {
        val whiteKing = Piece(Type.KING, Color.WHITE)
        assertTrue(whiteKing == whiteKing)
        assertFalse(whiteKing == Any())
        assertTrue(whiteKing == Piece(Type.KING, Color.WHITE))
    }

    @Test
    fun hashCodeTest() {
        val whiteKing = Piece(Type.KING, Color.WHITE)
        assertNotEquals(whiteKing.hashCode(), Any().hashCode())
        assertEquals(whiteKing.hashCode(), Piece(Type.KING, Color.WHITE).hashCode())
    }
}